TensorIterator: Check for memory overlap in all `nullary_op`s

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43423 Check for internal memory overlap in some indexing-type functions
* #43422 TensorIteratorConfig: Check memory overlap by default
* **#43421 TensorIterator: Check for memory overlap in all `nullary_op`s**
* #43420 TensorIterator: Check for memory overlap in all `compare_op`s
* #43419 TensorIterator: Check for memory overlap in all `binary_op`s
* #43418 TensorIterator: Check memory overlap in all `unary_op`s

Differential Revision: [D23298654](https://our.internmc.facebook.com/intern/diff/D23298654)

peterbell10 committed:
TensorIterator: Check for memory overlap in all `nullary_op`s

[ghstack-poisoned]

peterbell10 committed:
Update on "TensorIterator: Check for memory overlap in all `nullary_op`s"




[ghstack-poisoned]

zou3519 commented on test/test_torch.py:
```diff
@@ -14185,6 +14181,22 @@ def test_atanh_domain_float(self, device, dtype):
+            ("log_normal_", ()),
+            ("exponential_", ()),
+            ("geometric_", (0.5,)),
+            ("normal_", ()),
```
We should add a test for bernoulli_ too: https://github.com/pytorch/pytorch/blob/284ff047927e3b7f44f17a28f62c796bca93cea3/aten/src/ATen/native/cuda/DistributionBernoulli.cu#L38
https://github.com/pytorch/pytorch/blob/284ff047927e3b7f44f17a28f62c796bca93cea3/aten/src/ATen/native/cpu/DistributionTemplates.h#L345

Some versions of bernoulli go through different code paths :/ we should probably try to keep them consistent in a single PR: https://github.com/pytorch/pytorch/blob/284ff047927e3b7f44f17a28f62c796bca93cea3/aten/src/ATen/native/cpu/DistributionTemplates.h#L318-L322

zou3519 replied:
(see https://github.com/pytorch/pytorch/pull/43421#issuecomment-679386089)

zou3519 reviewed (commented):
I went through all use cases of nullary_op. Those look fine to me; we should add a test case for bernoulli_

zou3519 commented:
Hmm, actually, I see the bernoulli_ test code in the next diff up. Is that because it is only in the next diff that we finish moving all of the bernoulli code paths to checking memory overlap? If so then I think it is OK to leave it there.

I am probably going to try to land #43418 - #43421 together after this stack gets rebased and I'll take a look at #43422 and #43423 tomorrow

zou3519 reviewed (approved):
pre-emptive approve b/c the changes remaining are not that big

peterbell10 commented:
>Hmm, actually, I see the bernoulli_ test code in the next diff up. Is that because it is only in the next diff that we finish moving all of the bernoulli code paths to checking memory overlap?

There are different code paths depending on if MKL is used. One goes through `TensorIterator` and is checked in this PR, the other doesn't and has a `assert_no_internal_overlap` added manually in the next PR. I probably shouldn't have split it up that way, but it was difficult to see which tests belonged with each change.

zou3519 commented:
> There are different code paths depending on if the probability is a tensor. One goes through nullary_op and is checked in this PR, the other doesn't

Okay, I see. This is fine as-is then

peterbell10 committed:
Update on "TensorIterator: Check for memory overlap in all `nullary_op`s"


Differential Revision: [D23298654](https://our.internmc.facebook.com/intern/diff/D23298654)

[ghstack-poisoned]

peterbell10 committed:
Update on "TensorIterator: Check for memory overlap in all `nullary_op`s"


Differential Revision: [D23298654](https://our.internmc.facebook.com/intern/diff/D23298654)

[ghstack-poisoned]

peterbell10 committed:
Update on "TensorIterator: Check for memory overlap in all `nullary_op`s"


Differential Revision: [D23298654](https://our.internmc.facebook.com/intern/diff/D23298654)

[ghstack-poisoned]

peterbell10 committed:
Update on "TensorIterator: Check for memory overlap in all `nullary_op`s"


Differential Revision: [D23298654](https://our.internmc.facebook.com/intern/diff/D23298654)

[ghstack-poisoned]
